elasticapm/utils/cloud.py

```diff
@@ -0,0 +1,161 @@
+#  BSD 3-Clause License
+#
+#  Copyright (c) 2019, Elasticsearch BV
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import os
+import socket
+
+import urllib3
+
+
+def aws_metadata():
+    """
+    Fetch AWS metadata from the local metadata server. If metadata server is
+    not found, return an empty dictionary
+    """
+    ret = {}
+    http = urllib3.PoolManager()
+
+    try:
+        # This will throw an error if the metadata server isn't available,
+        # and will be quiet in the logs, unlike urllib3
+        socket.create_connection(("169.254.169.254", 80), 0.1)
+
+        ttl_header = {"X-aws-ec2-metadata-token-ttl-seconds": "300"}
+        token_url = "http://169.254.169.254/latest/api/token"
+        token_request = http.request("PUT", token_url, headers=ttl_header, timeout=3.0)
+        token = token_request.data.decode("utf-8")
+        aws_token_header = {"X-aws-ec2-metadata-token": token} if token else {}
```

Comment:
Interestingly, on the test hosts that I've used so far, you don't actually need a token. It's only relevant for [IMDSv2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html). Note that even on v1 hosts the code works fine. Since there's no actual credentials involved, I don't think leaving breadcrumbs for the user is useful, I'm not sure what they would actually do.... but I supposed we could do something at a low log level, like debug.

Reply:
Hmm, I'm actually not sure how to detect if the failure was due to token issues. Every AWS host I've spun up is IMDSv1 so it's not looking for a token...